lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp

```diff
@@ -115,77 +117,77 @@ enum ConvolutionKernelDimension {
   INVALID_KERNEL_DIM = -3
 };
 
-static tensor::EmptyOp generateTransposeDPSOutput(Value input, int64_t dim0,
-                                                  int64_t dim1,
-                                                  PatternRewriter &rewriter) {
-  auto input_type = mlir::cast<mlir::RankedTensorType>(input.getType());
-  auto output_shape = input_type.getShape().vec();
-  std::swap(output_shape[dim0], output_shape[dim1]);
-
-  auto output_type = RankedTensorType::get(
-      output_shape, input_type.getElementType(), input_type.getEncoding());
+/*
+ * Generates a sequence of dims in which to transpose to make current_layout
+ * match desired_layout
+ *
+ * Ex: if current_layout = [0, 1, 2, 3] and desired_layout = [0, 2, 3, 1]
+ * then the function will return [(1, 2), (2, 3)] because when we swap
+ * current_layout[1] with current_layout[2] we get [0, 2, 1, 3], and then when
+ * we swap current_layout[2] with current_layout[3] we get [0, 2, 3, 1], which
+ * is the desired layout
+ */
+static std::vector<TransposeDims>
+generateTransposeIndices(std::vector<int64_t> current_layout,
+                         const std::vector<int64_t> desired_layout) {
+  std::vector<TransposeDims> transpose_indices;
+  for (int64_t i = 0; i < static_cast<int64_t>(current_layout.size()); i++) {
+    if (current_layout[i] != desired_layout[i]) {
+      int64_t dim0 = i;
+      int64_t dim1 = std::find(current_layout.begin(), current_layout.end(),
+                               desired_layout[i]) -
+                     current_layout.begin();
+      transpose_indices.push_back(std::make_tuple(dim0, dim1));
+      std::swap(current_layout[dim0], current_layout[dim1]);
+    }
+  }
 
-  return rewriter.create<tensor::EmptyOp>(input.getLoc(), output_shape,
-                                          output_type.getElementType());
+  return transpose_indices;
 }
 
-static ttir::TransposeOp
-generateTranspose(Value input, int64_t dim0, int64_t dim1,
-                  PatternRewriter &rewriter,
-                  ::mlir::ArrayAttr operandConstraints) {
-  auto input_type = mlir::cast<mlir::RankedTensorType>(input.getType());
-  auto output_shape = input_type.getShape().vec();
-  std::swap(output_shape[dim0], output_shape[dim1]);
-
-  auto dim0_attr = rewriter.getSI32IntegerAttr(dim0);
-  auto dim1_attr = rewriter.getSI32IntegerAttr(dim1);
-
-  auto dps_output = generateTransposeDPSOutput(input, dim0, dim1, rewriter);
-  return rewriter.create<ttir::TransposeOp>(
-      input.getLoc(), dps_output.getType(), input, dps_output, dim0_attr,
-      dim1_attr, operandConstraints);
-}
+/*
+ * This function will use a sequence of transpose indices to
+ * generate the actual transpose operations descrbibed by them.
+ *
+ * It takes an input to apply these transposes to and returns the
+ * result at the end of the sequence
+ */
+static Value
+generateTransposeSequence(Value input, PatternRewriter &rewriter,
```

Comment:
Consider renaming `generateTransposeSequence` to `applyTransposeIndices`.

Reply:
I think `generateTransposeOps` might be better